src/platforms/esp32/components/avm_sys/sys.c

```diff
@@ -759,3 +785,76 @@ term esp_err_to_term(GlobalContext *glb, esp_err_t status)
             return term_from_int(status);
     }
 }
+
+int sys_mbedtls_entropy_func(void *entropy, unsigned char *buf, size_t size)
+{
+#ifndef MBEDTLS_THREADING_C
+    struct ESP32PlatformData *platform
+        = CONTAINER_OF(entropy, struct ESP32PlatformData, entropy_ctx);
+    SMP_MUTEX_LOCK(platform->entropy_mutex);
+    int result = mbedtls_entropy_func(entropy, buf, size);
+    SMP_MUTEX_UNLOCK(platform->entropy_mutex);
+
+    return result;
+#else
+    return mbedtls_entropy_func(entropy, buf, size);
+#endif
+}
+
+mbedtls_entropy_context *sys_mbedtls_get_entropy_context_lock(GlobalContext *global)
+{
+    struct ESP32PlatformData *platform = global->platform_data;
+
+    SMP_MUTEX_LOCK(platform->entropy_mutex);
+
+    if (!platform->entropy_is_initialized) {
+        // mbedtls_entropy_init will gather entropy for us using esp_fill_random()
+        // however it will be pseudorandom when wifi/bluetooth/RF is not enabled.
+        //
+        // TODO: when radio is not active bootloader_random_enable() must be enabled for gathering
+        // entropy. However, "enable function is not safe to use if any other subsystem is
+        // accessing the RF subsystem or the ADC at the same time".
+        mbedtls_entropy_init(&platform->entropy_ctx);
+        platform->entropy_is_initialized = true;
+    }
+
+    return &platform->entropy_ctx;
+}
+
+void sys_mbedtls_entropy_context_unlock(GlobalContext *global)
+{
+    struct ESP32PlatformData *platform = global->platform_data;
+    SMP_MUTEX_UNLOCK(platform->entropy_mutex);
+}
+
+mbedtls_ctr_drbg_context *sys_mbedtls_get_ctr_drbg_context_lock(GlobalContext *global)
+{
+    struct ESP32PlatformData *platform = global->platform_data;
+
+    SMP_MUTEX_LOCK(platform->random_mutex);
+
+    if (!platform->random_is_initialized) {
+        mbedtls_ctr_drbg_init(&platform->random_ctx);
+
+        mbedtls_entropy_context *entropy_ctx = sys_mbedtls_get_entropy_context_lock(global);
```

Comment:
I am still not a fan of this.

If would deadlock if first calling code first gets entropy and then ctr_drbg.